### PR TITLE
support log rotate by size

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**

### CODE REVIEW

File: src/util/file_log.rs [hunk 1]
```diff
@@ -57,47 +57,92 @@ fn open_log_file(path: &str) -> io::Result<File> {
...
+                    &time::strftime("%Y%m%d", &one_day_before(self.rotate_time.unwrap())).unwrap(),
+                );
+                self.update_rotate_time();
+            }
+            RotateType::BySize => {
+                let now = match now {
+                    Some(now) => now,
+                    None => time::now(),
+                };
+                s.push_str(&time::strftime("%Y%m%d%H%M%S", &now).unwrap());
```
Comment (siddontang) [inline src/util/file_log.rs]: can we use a separator to split the date and time, it is difficult to check the file name now. 

IMO, the format "%Y%m%d" or  "%Y%m%d%H%M%S" is compact and not intuitional.

Comment (huachaohuang) [reply src/util/file_log.rs]: I think we should agree on a format for tidb/tikv/pd.

Comment (zhangjinpeng1987) [reply src/util/file_log.rs]: Look at `tidb-2018-02-01T10-56-28.376.log`, so ugly - -

File: src/config.rs [hunk 2]
```diff
@@ -690,6 +691,7 @@ impl Default for TiKvConfig {
         TiKvConfig {
             log_level: LogLevelFilter::Info,
             log_file: "".to_owned(),
+            log_rotate_size: ReadableSize::mb(0),
```
Comment (overvenus) [inline src/config.rs]: Please check the `log_rotate_size` in the `validate`, in case it rotates too fast.

File: src/util/file_log.rs [hunk 3]
```diff
@@ -57,47 +57,92 @@ fn open_log_file(path: &str) -> io::Result<File> {
...
+#[derive(Clone, Copy)]
+enum RotateType {
+    ByDay,
+    BySize,
+}
+
 struct RotatingFileLoggerCore {
-    rollover_time: Tm,
+    rotate_type: RotateType,
+    rotate_size: Option<usize>,
```
Comment (overvenus) [inline src/util/file_log.rs]: How about put the `rotate_size` in the `RotateType::BySize(usize)`?

Comment (huachaohuang) [reply src/util/file_log.rs]: And `RotateType::ByDay(Tm)`.

File: src/util/file_log.rs (hunk 1 above)
Comment (siddontang) [reply src/util/file_log.rs]: For simplicity, I think 20180201_105628.log is fine.

**Comment by zhangjinpeng1987** [comment]
@overvenus @huachaohuang PTAL

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -96,13 +96,14 @@ fn init_log(config: &TiKvConfig) {
...
         });
     } else {
-        let w = RotatingFileLogger::new(&config.log_file).unwrap_or_else(|e| {
-            fatal!(
-                "failed to initial log with file {:?}: {:?}",
-                config.log_file,
-                e
-            );
-        });
+        let w = RotatingFileLogger::new(&config.log_file, Some(config.log_rotate_size.0 as usize))
```
Comment (huachaohuang) [inline src/bin/tikv-server.rs]: Should be `None` when `config.log_rotate_size == 0`?

File: src/config.rs [hunk 5]
```diff
@@ -779,6 +782,12 @@ impl TiKvConfig {
         if !db_exist(&kv_db_path) && db_exist(&self.raft_store.raftdb_path) {
             return Err("default rocksdb not exist, buf raftdb exist".into());
         }
+        if self.log_rotate_size.0 > 0 && self.log_rotate_size.0 < LOG_MIN_SIZE {
```
Comment (Hoverbear) [inline src/config.rs]: Is it possible to remove `self.log_rotate_size.0 > 0 &&`?

**Comment by huachaohuang** [comment]
Since we are planning to use slog for this, I think we can close it now.

**Pull request closed without merging**